src/hooks/useCache.js

```diff
@@ -0,0 +1,111 @@
+import { has, isEqual, isPlainObject, pick } from 'lodash';
+import { useRef, useMemo, useEffect, useCallback } from 'react';
+import { useSelector } from 'react-redux';
+
+const document = ({ databaseOverrides = {}, database = {} }, [path, id]) => {
+  const override = databaseOverrides[path] && databaseOverrides[path][id];
+  const doc = database[path] && database[path][id];
+  if (override) return { ...doc, ...override };
+  return doc;
+};
+
+const selectDocument = (cache, id, path, fields = null) => {
+  if (id === undefined || path === undefined) return undefined;
+  const doc = document(cache, [path, id]);
+  if (!doc) return doc;
+  if (typeof fields === 'string') {
+    return doc[fields];
+  }
+  if (doc && fields) {
+    return pick(doc, ['id', 'path', ...fields]);
+  }
+  return doc;
+};
+
+const selectList = (cache, result, fields = null) => {
+  if (!result || !result.ordered) return undefined;
+  const docs = result.ordered.reduce((arr, [path, id]) => {
+    const doc = selectDocument(cache, id, path, fields);
+    if (doc) {
+      arr.push(doc);
+    }
+    return arr;
+  }, []);
+
+  return result.id && Array.isArray(docs) ? docs[0] : docs;
+};
+
+const selectAlias = (state, alias) =>
+  (state &&
+    state.firestore &&
+    state.firestore.cache &&
+    state.firestore.cache[alias]) ||
+  undefined;
+
+/**
+ * set/uset listeners and return a selector to it.
+ * Note: functions are supported but don't use them.
```

Comment:
Don't fully understand this comment, nor why not to use it.

Reply:
Good question. So to see the problem with functions look at redux's [createSelector](https://redux.js.org/usage/deriving-data-selectors) function (also see [idiomatic redux optimizing for performance]((https://blog.isquaredsoftware.com/2017/12/idiomatic-redux-using-reselect-selectors/#optimizing-performance-with-reselect))) and our [compute sprint effort](https://github.com/TARAAI/tara-js/blob/development/microservices/app/src/reduxStore/sprints/selectors.ts) functions. **It is so complex you are more likely it get it wrong than right.**

The core need was coders want to combine local props data with redux data. The flaw was that the library authors let you use a dynamic function which captures local scope. Look at our issue with useCallback dependencies. It spirals so quickly into many pitfalls and edge cases that it requires too much custom memoization and for normal coder to really get it right 100% of the time. 

The solution is a simpler API that allows the user to do the right thing 100% of the time (without sacrificing performance). So instead we only let users pass a string of the values to select. Then they can use a normal function to combine with local data. 

The useRead will only trigger when it's data will actually change. And the local prop data will also trigger when it's data as changed. So when the component needs to re-render it's because the source of a computed field has change it will actually trigger a render. This forces simplification and ensure the coder can't do stupid things from an overly complicated API.